cpp/src/arrow/dataset/partition_test.cc

```diff
@@ -1107,5 +1108,88 @@ TEST(TestStripPrefixAndFilename, Basic) {
                                           "year=2019/month=12/day=01"));
 }
 
+TEST_F(TestPartitioning, HivePartitionReadEvaluation) {
```

Comment:
This test is too complex.  Could you create a narrower unit test that focuses on the needed behavior?  For example, there should be no need to create a filesystem and actually write files if your goal is to test the partitioning factory.  The partitioning factory takes strings & schemas as input and outputs a partitioning object.

For example, you can reproduce the issue with this:

```diff
diff --git a/cpp/src/arrow/dataset/partition_test.cc b/cpp/src/arrow/dataset/partition_test.cc
index 3e681a9cb..0e22152e0 100644
--- a/cpp/src/arrow/dataset/partition_test.cc
+++ b/cpp/src/arrow/dataset/partition_test.cc
@@ -83,8 +83,14 @@ class TestPartitioning : public ::testing::Test {
   void AssertInspect(const std::vector<std::string>& paths,
                      const std::vector<std::shared_ptr<Field>>& expected) {
     ASSERT_OK_AND_ASSIGN(auto actual, factory_->Inspect(paths));
-    ASSERT_EQ(*actual, Schema(expected));
-    ASSERT_OK_AND_ASSIGN(partitioning_, factory_->Finish(actual));
+    Schema expected_schema(expected);
+    ASSERT_EQ(*actual, expected_schema);
+    // The dataset will usually have more than just partition fields.  The partitioning
+    // schema should ignore these.
+    std::vector<std::shared_ptr<Field>> all_fields(expected);
+    all_fields.push_back(field("field_unrelated_to_partitioning", int8()));
+    ASSERT_OK_AND_ASSIGN(partitioning_, factory_->Finish(schema(std::move(all_fields))));
+    AssertSchemaEqual(expected_schema, *partitioning_->schema());
   }
 
   void AssertPartition(const std::shared_ptr<Partitioning> partitioning,
@@ -571,6 +577,11 @@ TEST_F(TestPartitioning, DiscoverHiveSchema) {
   // missing path segments will not cause an error
   AssertInspect({"/alpha=0/beta=1", "/beta=2/alpha=3", "/gamma=what"},
                 {Int("alpha"), Int("beta"), Str("gamma")});
+
+  // No partitioning information should yield an empty partitioning even if
+  // there are fields in the dataset schema
+  AssertInspect({"/foo/chunk.parquet", "/bar/chunk.parquet", "/baz"}, {});
+  AssertInspect({"chunk.parquet"}, {});
 }
 
 TEST_F(TestPartitioning, HiveDictionaryInference) {
```

Reply:
I removed the added test case and updated with your suggestion. I understand your concern.

Reply:
should we include another test case? Or this change would be satisfactory?

Reply:
This is fine.  It covers your change.  Are there other branches you feel we are not testing?

Reply:
Not quite. But I felt the need for a roundtrip test. Other than that, I don’t have strong concerns.

Reply:
We have quite a few roundtrip tests for dataset reading and writing already (python's test_dataset has quite a few) so I'd be opposed to another one unless there was a specific reason.